### PR TITLE
fix(cheatcodes): fail when trying to parse malformatted strings as addresses

### LINKED ISSUE #8778: bug(`cheatcodes`): `vm.parseJson()`: invalid address field will get packed 0s
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0

### What command(s) is the bug in?

vm.parseJson()

### Operating System

macOS (Apple Silicon)

### Describe the bug

If a json contains an invalid address (for example, an address that is 39 characters long instead of 40), foundry will add an extra 0 onto the address so that it can be packed into an `address` object. 

Ex: 
- "address" in json `0x242ba6d68FfEb4a098B591B32d370F973FF882B`
- `address` object parsed by jsonParse => `0x0242ba6d68FfEb4a098B591B32d370F973FF882B`
 
This seems unintended and can lead to critical issues like sending funds/tokens to unowned addresses. At the very least, foundry should revert/error if this case is found. Would it be possible to byte-length validation on expected types?

### PR DESCRIPTION
**Pull request by fmorisan**
Hi, I'm here trying to solve #8778.

This is my first time contributing here, and also my first time writing Rust on a real project. Please let me know if the approach I took is wrong here. Thanks!

## Motivation
Fix #8778 as it seemed like a small issue I could take on, and the current behaviour bugged me today as well.

## Solution
Use the already imported `alloy_dyn_abi` crate to try and parse addresses when the passed in `DynSolType` is an address. We do some minimal validation on the passed in string, and handle it with the underlying crate.

### CODE REVIEW

**Review by klkvr — CHANGES_REQUESTED** [review]
The reported issues is about the `parseJson` cheatcode, while your fix is related to typed parsing (e.g. `parseJsonAddress`), which already performs validation in `string::parse_value`. Additionaly, this will break parsing for non-checksummed addresses which is not something we want.

The issue is caused by this code: https://github.com/foundry-rs/foundry/blob/818eeb9d5018d3858238d925fa9c9ef5fcdaee47/crates/cheatcodes/src/json.rs#L572-L575

I think we should not add `0` if string length is 39, and instead throw a error asking user to be explicit on whether this is supposed to be an `address` or `bytes`

**Comment by fmorisan** [comment]
Alright so we're basically guessing on the underlying type from the string length. Pushed the requested change.

I encountered this issue while using `vm.parseJsonType` as well, though I might have been using an earlier `foundry` version at the time.

File: crates/cheatcodes/src/json.rs [hunk 1]
```diff
@@ -568,7 +568,10 @@ pub(super) fn json_value_to_token(value: &Value) -> Result<DynSolValue> {
         }
         Value::String(string) => {
             if let Some(mut val) = string.strip_prefix("0x") {
-                let s;
+                let s; 
+                if val.len() == 39 {
+                    return Err(format!("Cannot parse \"{val}\" as an address. If you want to specify bytes, prepend zeroes to the value.").into())
```
Comment (klkvr) [inline crates/cheatcodes/src/json.rs:573]:
```suggestion
                    return Err(format!("Cannot parse \"{val}\" as an address. If you want to specify address, prepend zero to the value.").into())
```

**Review by klkvr — APPROVED** [review]

**Comment by fmorisan** [comment]
accepted the error message suggestion, and rebased to `main` branch tip. thanks!

**Pull request merged**